src/NuGetPackageVerifier/Rules/PackageVersionMatchesAssemblyVersionRule.cs

```diff
@@ -6,40 +6,86 @@
 using Mono.Cecil;
+using NuGet.Versioning;
 
 namespace NuGetPackageVerifier.Rules
 {
     public class PackageVersionMatchesAssemblyVersionRule : AssemblyHasAttributeRuleBase
     {
-        private Version _packageVersion;
+        private NuGetVersion _packageVersion;
         private string _packageId;
 
         public override IEnumerable<PackageVerifierIssue> Validate(PackageAnalysisContext context)
         {
-            _packageVersion = context.Metadata.Version.Version;
+            _packageVersion = context.Metadata.Version;
             _packageId = context.Metadata.Id;
             return base.Validate(context);
         }
 
         public override IEnumerable<PackageVerifierIssue> ValidateAttribute(string currentFilePath,
             Mono.Collections.Generic.Collection<CustomAttribute> assemblyAttributes)
         {
-            var versionAttribute = assemblyAttributes.SingleOrDefault(a =>
+            var assemblyInformationalVersionAttribute = assemblyAttributes.SingleOrDefault(a =>
                 a.AttributeType.FullName.Equals(
-                    typeof(AssemblyInformationalVersionAttribute).FullName ??
-                    typeof(AssemblyVersionAttribute).FullName,
+                    typeof(AssemblyInformationalVersionAttribute).FullName,
                     StringComparison.Ordinal));
 
-            if (versionAttribute == null)
+            var assemblyInformationalNuGetVersion = new NuGetVersion(assemblyInformationalVersionAttribute.ConstructorArguments[0].Value.ToString());
+            if (!VersionEquals(_packageVersion, assemblyInformationalNuGetVersion))
             {
-                yield break;
+                yield return PackageIssueFactory.AssemblyInformationalVersionDoesNotMatchPackageVersion(
+                    currentFilePath,
+                    assemblyInformationalNuGetVersion,
+                    _packageVersion,
+                    _packageId);
             }
 
-            var assemblyVersion = versionAttribute.AttributeType.Module.Assembly.Name.Version;
-            if (!_packageVersion.Equals(assemblyVersion))
+            var assemblyFileVersionAttribute = assemblyAttributes.SingleOrDefault(a =>
+                a.AttributeType.FullName.Equals(
+                    typeof(AssemblyFileVersionAttribute).FullName,
+                    StringComparison.Ordinal));
+
+            var assemblyFileNuGetVersion = new NuGetVersion(assemblyFileVersionAttribute.ConstructorArguments[0].Value.ToString());
+            if (!VersionEquals(_packageVersion, assemblyFileNuGetVersion))
+            {
+                yield return PackageIssueFactory.AssemblyFileVersionDoesNotMatchPackageVersion(
+                    currentFilePath,
+                    assemblyFileNuGetVersion,
+                    _packageVersion,
+                    _packageId);
+            }
+
+            var assemblyVersion = assemblyInformationalVersionAttribute.AttributeType.Module.Assembly.Name.Version.ToString();
```

Comment:
Hmm this seems like an odd way to get the assembly version. That should presumably be directly available from the assembly itself, no? Though, this code still looks a bit off to me -- is this getting the assembly version of the assembly that defines AssemblyInformationalVersionAttribute, which is mscorlib? Or is it getting the assembly version of the assembly we're testing?

Either way, I'm suspecting that the rule base class that this rule derives from is just lacking a few things that would make writing this code easier...